##1 Résumé de la mission

Data Analyst freelance  en mission chez BottleNeck, un marchand de vin prestigieux.

1. de rapprocher un ERP des références produit avec une table issue de l'outil cms contenant les informations à propos de ces produits
2. calculer le chiffre d'affaires par produit
3. calculer le total du chiffre d'affaire réalisé en ligne
4. rechercher d'éventuelles valeurs aberrantes dans les prix des produits, et faire une liste et une représentation graphique 

In [122]:
import pandas as pd 
import numpy as np
import plotly.express as px

In [123]:
erp = pd.read_excel('erp.xlsx')
liaison = pd.read_excel('liaison.xlsx')
web = pd.read_excel('web.xlsx')


/usr/local/lib/python3.7/dist-packages/openpyxl/worksheet/_reader.py:312: UserWarning:

Unknown extension is not supported and will be removed



In [124]:
 print(web.shape)
web.head()

(1513, 28)


sku  virtual  downloadable  rating_count  average_rating  \
0  bon-cadeau-25-euros        0             0             0             0.0   
1                15298        0             0             0             0.0   
2                15296        0             0             0             0.0   
3                15300        0             0             0             0.0   
4                19814        0             0             0             0.0   

   total_sales tax_status  tax_class  post_author           post_date  ...  \
0         10.0    taxable        NaN          1.0 2018-06-01 13:53:46  ...   
1          6.0    taxable        NaN          2.0 2018-02-08 12:58:52  ...   
2          0.0    taxable        NaN          2.0 2018-02-08 13:49:41  ...   
3          0.0    taxable        NaN          2.0 2018-02-08 14:08:36  ...   
4          3.0    taxable        NaN          2.0 2018-02-09 14:01:05  ...   

                                           post_name       post_modified  \
0                             bon-cadeau-de-25-euros 2018-06-01 14:13:57   
1        pierre-jean-villa-saint-joseph-preface-2018 2019-12-30 09:30:29   
2          pierre-jean-villa-saint-joseph-tilde-2017 2019-12-21 09:00:17   
3  pierre-jean-villa-croze-hermitage-accroche-coe... 2020-06-26 18:15:03   
4                  pierre-jean-villa-igp-gamine-2018 2020-01-04 16:36:01   

    post_modified_gmt post_content_filtered post_parent  \
0 2018-06-01 12:13:57                   NaN         0.0   
1 2019-12-30 08:30:29                   NaN         0.0   
2 2019-12-21 08:00:17                   NaN         0.0   
3 2020-06-26 16:15:03                   NaN         0.0   
4 2020-01-04 15:36:01                   NaN         0.0   

                                                guid menu_order  post_type  \
0  https://www.bottle-neck.fr/?post_type=product&...        0.0    product   
1  https://www.bottle-neck.fr/?post_type=product&...        0.0    product   
2  https://www.bottle-neck.fr/?post_type=product&...        0.0    product   
3  https://www.bottle-neck.fr/?post_type=product&...        0.0    product   
4  https://www.bottle-neck.fr/?post_type=product&...        0.0    product   

  post_mime_type comment_count  
0            NaN           0.0  
1            NaN           0.0  
2            NaN           0.0  
3            NaN           0.0  
4            NaN           0.0  

[5 rows x 28 columns]

In [125]:
web.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1513 entries, 0 to 1512
Data columns (total 28 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   sku                    1428 non-null   object        
 1   virtual                1513 non-null   int64         
 2   downloadable           1513 non-null   int64         
 3   rating_count           1513 non-null   int64         
 4   average_rating         1430 non-null   float64       
 5   total_sales            1430 non-null   float64       
 6   tax_status             716 non-null    object        
 7   tax_class              0 non-null      float64       
 8   post_author            1430 non-null   float64       
 9   post_date              1430 non-null   datetime64[ns]
 10  post_date_gmt          1430 non-null   datetime64[ns]
 11  post_content           0 non-null      float64       
 12  post_title             1430 non-null   object        
 13  pos

In [126]:
web.isna().sum()

sku                        85
virtual                     0
downloadable                0
rating_count                0
average_rating             83
total_sales                83
tax_status                797
tax_class                1513
post_author                83
post_date                  83
post_date_gmt              83
post_content             1513
post_title                 83
post_excerpt              797
post_status                83
comment_status             83
ping_status                83
post_password            1513
post_name                  83
post_modified              83
post_modified_gmt          83
post_content_filtered    1513
post_parent                83
guid                       83
menu_order                 83
post_type                  83
post_mime_type            799
comment_count              83
dtype: int64

On dénote 85 valeurs nulles

In [127]:
print(liaison.shape)
liaison.head()

(825, 2)


product_id id_web
0        3847  15298
1        3849  15296
2        3850  15300
3        4032  19814
4        4039  19815

In [128]:
liaison.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 825 entries, 0 to 824
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   product_id  825 non-null    int64 
 1   id_web      734 non-null    object
dtypes: int64(1), object(1)
memory usage: 13.0+ KB


In [129]:
print(erp.shape)
erp.head()

(825, 5)


product_id  onsale_web  price  stock_quantity stock_status
0        3847           1   24.2               0   outofstock
1        3849           1   34.3               0   outofstock
2        3850           1   20.8               0   outofstock
3        4032           1   14.1               0   outofstock
4        4039           1   46.0               0   outofstock

In [130]:
erp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 825 entries, 0 to 824
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   product_id      825 non-null    int64  
 1   onsale_web      825 non-null    int64  
 2   price           825 non-null    float64
 3   stock_quantity  825 non-null    int64  
 4   stock_status    825 non-null    object 
dtypes: float64(1), int64(3), object(1)
memory usage: 32.4+ KB


# On commence par compter le nombre de produits uniques référencés dans la DF

In [131]:
web.sku.nunique()

714

714 références uniques, pour 825 références attendues cela signifie que l'ensemble des produits listés dans les DF erp et liaison ne sont pas commercialisés en ligne

In [132]:
webunique = web[web['sku'] == 15298]
webunique

sku  virtual  downloadable  rating_count  average_rating  total_sales  \
1    15298        0             0             0             0.0          6.0   
799  15298        0             0             0             0.0          6.0   

    tax_status  tax_class  post_author           post_date  ...  \
1      taxable        NaN          2.0 2018-02-08 12:58:52  ...   
799        NaN        NaN          2.0 2018-02-08 12:58:52  ...   

                                       post_name       post_modified  \
1    pierre-jean-villa-saint-joseph-preface-2018 2019-12-30 09:30:29   
799  pierre-jean-villa-saint-joseph-preface-2018 2019-12-30 09:30:29   

      post_modified_gmt post_content_filtered post_parent  \
1   2019-12-30 08:30:29                   NaN         0.0   
799 2019-12-30 08:30:29                   NaN         0.0   

                                                  guid menu_order   post_type  \
1    https://www.bottle-neck.fr/?post_type=product&...        0.0     product   
799  https://www.bottle-neck.fr/wp-content/uploads/...        0.0  attachment   

    post_mime_type comment_count  
1              NaN           0.0  
799     image/jpeg           0.0  

[2 rows x 28 columns]

La DF web est issue du Content Management System du site web.

Nous constatons que les produits sont référencés dans la DF web au travers de deux lignes.

en observant la colonne post_type, on comprend que l'une des deux lignes correspond au texte descriptif du produit quand l'autre correspond aux illustrations du produit sur le site.

c'est donc une colonne adéquate pour éliminer les doublons en ne gardant qu'un seul type de ligne.

Nous conserverons donc les colonnes:

1. "sku" que nous renommerons en "id_web" pour faciliter la lecture et la jointure avec la df liaison.

2. "total_sales" dont nous aurons besoin pour les calculs de CA qui nous ont été demandés.

3. "post_name" que nous renommerons en "product_name" pour connaitre le nom des produits

4. "post_type" afin de pouvoir trier les doublons


# nettoyage et mise en forme des données df web

# changement nom de colonnes

In [133]:
web.rename(columns={"sku": "id_web","post_name" : "product_name"}, inplace=True)

# drop de colonnes

In [134]:
web = web.drop(["virtual", "downloadable", "rating_count","average_rating",  "tax_class", "post_author", "post_date","post_date_gmt", "post_content", "post_title", "post_excerpt", "post_status", "comment_status", "ping_status", "post_password", "post_modified", "post_modified_gmt", "post_content_filtered", "post_parent", "guid", "menu_order", "tax_status", "post_mime_type", "comment_count" ], axis=1)
erp = erp.drop(["stock_status"], axis=1)

# drop de lignes

In [135]:
web = web[web.post_type == 'product']

In [136]:
print(web.shape)
web.head()

(716, 4)


id_web  total_sales  \
0  bon-cadeau-25-euros         10.0   
1                15298          6.0   
2                15296          0.0   
3                15300          0.0   
4                19814          3.0   

                                        product_name post_type  
0                             bon-cadeau-de-25-euros   product  
1        pierre-jean-villa-saint-joseph-preface-2018   product  
2          pierre-jean-villa-saint-joseph-tilde-2017   product  
3  pierre-jean-villa-croze-hermitage-accroche-coe...   product  
4                  pierre-jean-villa-igp-gamine-2018   product

Les questions posées correspondant exclusivement aux produits en ligne, nous éliminerons les produits non commercialisés sur le site.

créations de bases

In [137]:
all_bases = {"erp" : erp,"liaisons" : liaison,"web" : web}

 création d'une base erp_online qui ne contient que les lignes d'erp correspondant à des produits commercialisés en ligne.

In [138]:
erp_online = erp[erp.onsale_web == 1]

In [139]:
def cle_primaire(df_cle): # Vérification de la clé primaire
  table_length = len(df_cle) - len(df_cle.drop_duplicates())
  if table_length == 0:
    print('Clé primaire (0 doublon)')
  else:
    print('Pas une clé primaire (', table_length, 'doublons )')
    
cle_primaire(erp_online['product_id'])

Clé primaire (0 doublon)


In [140]:
def cle_primaire(df_cle): # Vérification de la clé primaire
  table_length = len(df_cle) - len(df_cle.drop_duplicates())
  if table_length == 0:
    print('Clé primaire (0 doublon)')
  else:
    print('Pas une clé primaire (', table_length, 'doublons )')
    
cle_primaire(liaison['product_id'])

Clé primaire (0 doublon)


 jointures
 entre l'erp "en ligne" et le fichier de liaison:

In [141]:
erp_online_liaison = pd.merge(erp_online, liaison, how="outer", on="product_id")

entre le fichier web et le fichier de liaison:

In [142]:
web_liaison = pd.merge(web, liaison, how="outer", on="id_web")

entre l'erp "en ligne" et la jointure du fichier web et liaison:

In [143]:
total_base = pd.merge(erp_online, web_liaison, on="product_id")

suppression des 3 NaN correspondant aux produits référencés dans erp_online et pas dans web

In [144]:
total_base = total_base[total_base.id_web.isna() == False]

calcul de la valeur du stock

In [145]:
total_base["valeur_stock"] = total_base.stock_quantity * total_base.price

chiffre d'affaire 

In [146]:
total_base["chiffre_affaire"] = total_base['price']*total_base['total_sales']

valeur stock > CA

In [147]:
total_base["stockSupCA"] = total_base.valeur_stock > total_base.chiffre_affaire

ajout d'un Booleen si le produit s'est déjà vendu

In [148]:
total_base["sold"] = total_base.chiffre_affaire > 0 

In [149]:
total_base.head()

product_id  onsale_web  price  stock_quantity id_web  total_sales  \
0        3847           1   24.2               0  15298          6.0   
1        3849           1   34.3               0  15296          0.0   
2        3850           1   20.8               0  15300          0.0   
3        4032           1   14.1               0  19814          3.0   
4        4039           1   46.0               0  19815          0.0   

                                        product_name post_type  valeur_stock  \
0        pierre-jean-villa-saint-joseph-preface-2018   product           0.0   
1          pierre-jean-villa-saint-joseph-tilde-2017   product           0.0   
2  pierre-jean-villa-croze-hermitage-accroche-coe...   product           0.0   
3                  pierre-jean-villa-igp-gamine-2018   product           0.0   
4          pierre-jean-villa-cote-rotie-carmina-2017   product           0.0   

   chiffre_affaire  stockSupCA   sold  
0            145.2       False   True  
1              0.0       False  False  
2              0.0       False  False  
3             42.3       False   True  
4              0.0       False  False

On avait créé une DF en procédant à une jointure entre les 3 DF fournies dans la section "variables et fonctions".

Résolution des problèmes énoncés
 On appelle une ligne au hazard pour avoir un apperçu des informations contenue dans la DF

In [150]:
total_base.iloc[230]

product_id                                                    4598
onsale_web                                                       1
price                                                         41.8
stock_quantity                                                  61
id_web                                                       15478
total_sales                                                    1.0
product_name       marc-colin-et-fils-saint-aubin-en-montceau-2017
post_type                                                  product
valeur_stock                                                2549.8
chiffre_affaire                                               41.8
stockSupCA                                                    True
sold                                                          True
Name: 232, dtype: object

Comme nous avions retiré de la DF tous les produits non commercialisés en ligne, il nous suffit de faire la somme des valeurs de la colonne chiffre_affaire

total du chiffre d'affaire réalisé en ligne

In [151]:
ca_total = total_base.chiffre_affaire.sum()
print(f"total du chiffre d'affaire en ligne : {ca_total} €")

total du chiffre d'affaire en ligne : 70568.6 €


le Chiffre d'affaire par produit est contenu dans la colonne chiffre_affaire pour chaque ligne.

chiffre d'affaire par produit classé par CA décroissant

In [152]:
total_base.sort_values("chiffre_affaire", ascending=False, inplace=True)
total_base.head()

product_id  onsale_web  price  stock_quantity id_web  total_sales  \
194        4334           1   49.0               0   7818         96.0   
71         4144           1   49.0              11   1662         87.0   
218        4402           1  176.0               8   3510         13.0   
70         4142           1   53.0               8  11641         30.0   
69         4141           1   39.0               1    304         40.0   

                              product_name post_type  valeur_stock  \
194  champagne-gosset-grand-blanc-de-blanc   product           0.0   
71             champagne-gosset-grand-rose   product         539.0   
218                   cognac-frapin-vip-xo   product        1408.0   
70   champagne-gosset-grand-millesime-2006   product         424.0   
69         gosset-champagne-grande-reserve   product          39.0   

     chiffre_affaire  stockSupCA  sold  
194           4704.0       False  True  
71            4263.0       False  True  
218           2288.0       False  True  
70            1590.0       False  True  
69            1560.0       False  True

statistiques rapides du chiffre d'affaire par produit

In [153]:
ca_positif = total_base[total_base.sold == True]
ca_positif.chiffre_affaire.describe()

count     385.000000
mean      183.295065
std       400.324073
min         6.500000
25%        38.600000
50%        81.600000
75%       164.400000
max      4704.000000
Name: chiffre_affaire, dtype: float64

La boutique en ligne à commercialisé  385 références sur 714.

La meilleure vente apporte 4704€ de CA, la moins bonne 6,50€. Le CA produit moyen est de 183.29€

# Recherche des valeurs anormales

Récupération des statistiques de la colonne price

In [154]:
total_base.price.describe(include="all")

count    714.000000
mean      32.493137
std       27.810525
min        5.200000
25%       14.100000
50%       23.550000
75%       42.175000
max      225.000000
Name: price, dtype: float64

In [155]:
data = total_base.price

In [156]:
q1 = np.percentile(data, 25)
q3 = np.percentile(data, 75)
interquartile = q3-q1
print(f'intervale interquartile : {interquartile}')

intervale interquartile : 28.075000000000003


Calcul minimum statistique 


In [157]:
min_stat= q1 - (interquartile*1.5)
print(f'minimum statistique :{min_stat}') 

minimum statistique :-28.012500000000003


In [158]:
max_stat=q3+(interquartile*1.5)
print(f'maximum statistique:{max_stat}')

maximum statistique:84.28750000000001


Recherche de valeurs dont le prix est supérieur au maximum statistique 

In [159]:
total_base_High = total_base[total_base['price'] > max_stat]
print(f'nombre de valeurs ayant des prix supérieurs au max statistique:{total_base_High.shape[0]}')

nombre de valeurs ayant des prix supérieurs au max statistique:32


Représentation graphique de la boite a moustache 

In [160]:
figu = px.box(data, x="price", 
points="all",
title=" boite à moustache répartition des prix produit")
figu.show()

Les outliers sont visibles sur la boite a moustache 

On va afficher les valeurs à contrôler 

In [161]:
total_base_High.sort_values("price", ascending=False, inplace=True)

/usr/local/lib/python3.7/dist-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



On va rechercher les ventes positives

In [162]:
total_base_highValue_isSold = total_base_High[(total_base_High.sold == 1)]
print(total_base_highValue_isSold.shape[0])
total_base_highValue_isSold.head(32)

10


product_id  onsale_web  price  stock_quantity id_web  total_sales  \
199        4352           1  225.0               0  15940          5.0   
593        5892           1  191.3              10  14983          3.0   
218        4402           1  176.0               8   3510         13.0   
383        4904           1  137.0              13  14220          5.0   
648        6126           1  135.0              10  14923          2.0   
201        4355           1  126.5               2  12589         11.0   
659        6212           1  115.0               2  13996          2.0   
219        4404           1  108.5               2   3507          2.0   
222        4407           1  104.0               6   3509          1.0   
205        4359           1   85.6               0  13853          1.0   

                                          product_name post_type  \
199     champagne-egly-ouriet-grand-cru-millesime-2008   product   
593  coteaux-champenois-egly-ouriet-ambonnay-rouge-...   product   
218                               cognac-frapin-vip-xo   product   
383  domaine-des-croix-corton-charlemagne-grand-cru...   product   
648             champagne-gosset-celebris-vintage-2007   product   
201  champagne-egly-ouriet-grand-cru-brut-blanc-de-...   product   
659  domaine-des-comtes-lafon-volnay-1er-cru-santen...   product   
219                         cognac-frapin-fontpinot-xo   product   
222                          cognac-frapin-cigar-blend   product   
205  champagne-larmandier-bernier-grand-cru-vieille...   product   

     valeur_stock  chiffre_affaire  stockSupCA  sold  
199           0.0           1125.0       False  True  
593        1913.0            573.9        True  True  
218        1408.0           2288.0       False  True  
383        1781.0            685.0        True  True  
648        1350.0            270.0        True  True  
201         253.0           1391.5       False  True  
659         230.0            230.0       False  True  
219         217.0            217.0       False  True  
222         624.0            104.0        True  True  
205           0.0             85.6       False  True

Après vérification rapide via  google, les prix de ces produits sont tout à fait dans la moyenne des tarifs proposés.

